File: frontend/js/views/pto_balance.js
// PTO balance statement generator — earned − used, capped, with payout value,
// via /calc/pto-balance.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const hrs = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderPtoBalance(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.ptobal.h1.title">// PTO BALANCE</span></h1>
        <p class="muted small" data-i18n="view.ptobal.hint.intro">
            A point-in-time statement of an employee's paid-time-off balance — distinct from the PTO policy.
            It computes hours earned over the periods worked, subtracts hours used, applies any accrual cap,
            and values the remaining balance for a separation payout. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.ptobal.h2.inputs">Balance inputs</h2>
            <form id="ptobal-form" class="inline-form">
                <label><span data-i18n="view.ptobal.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.ptobal.label.employer">Employer</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.ptobal.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.ptobal.label.rate">Accrual (hours/period)</span>
                    <input type="number" step="0.01" min="0" name="accrual_rate_hours_per_period" value="5" required></label>
                <label><span data-i18n="view.ptobal.label.periods">Periods worked</span>
                    <input type="number" step="1" min="0" name="periods_worked" value="20" required></label>
                <label><span data-i18n="view.ptobal.label.used">Hours used</span>
                    <input type="number" step="1" min="0" name="hours_used" value="30"></label>
                <label><span data-i18n="view.ptobal.label.cap">Accrual cap (hours, 0 = none)</span>
                    <input type="number" step="1" min="0" name="accrual_cap_hours" value="80"></label>
                <label><span data-i18n="view.ptobal.label.hourly">Hourly rate ($)</span>
                    <input type="number" step="0.01" min="0" name="hourly_rate_usd" value="25"></label>
                <label><span data-i18n="view.ptobal.label.hpd">Hours per workday</span>
                    <input type="number" step="0.5" min="1" name="hours_per_day" value="8"></label>
                <label><span data-i18n="view.ptobal.label.date">As-of date</span>
                    <input type="date" name="as_of_date" value="2026-07-01" required></label>
                <label><span data-i18n="view.ptobal.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.ptobal.ph.statute'))}"></label>
            </form>
        </div>
        <div id="ptobal-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#ptobal-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            employer_name: (fd.get('employer_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            accrual_rate_hours_per_period: Number(fd.get('accrual_rate_hours_per_period')) || 0,
            periods_worked: Number(fd.get('periods_worked')) || 0,
            hours_used: Number(fd.get('hours_used')) || 0,
            accrual_cap_hours: Number(fd.get('accrual_cap_hours')) || 0,
            hourly_rate_usd: Number(fd.get('hourly_rate_usd')) || 0,
            hours_per_day: Number(fd.get('hours_per_day')) || 8,
            as_of_date: fd.get('as_of_date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcPtoBalance(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.ptobal.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const capNote = doc.cap_applied
        ? ` <span class="muted small" data-i18n="view.ptobal.capped">(capped)</span>`
        : '';
    const el = mount.querySelector('#ptobal-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.ptobal.card.balance">Balance (hours)</div>
                    <div class="value">${hrs(doc.balance_hours)}${capNote}</div></div>
                <div class="card"><div class="label" data-i18n="view.ptobal.card.days">Balance (days)</div>
                    <div class="value">${hrs(doc.balance_days)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ptobal.card.earned">Earned / used</div>
                    <div class="value">${hrs(doc.hours_earned)} / ${hrs(doc.hours_used)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.ptobal.card.payout">Payout value</div>
                    <div class="value">${money(doc.payout_value_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="ptobal-copy" type="button" data-i18n="view.ptobal.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="ptobal-download" type="button" data-i18n="view.ptobal.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#ptobal-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.ptobal.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.ptobal.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#ptobal-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'pto-balance.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/pto_balance.rs
//! PTO balance statement — a point-in-time statement of an individual employee's
//! paid-time-off balance, distinct from the PTO policy (which states the accrual
//! rate). It computes hours earned over the periods worked, subtracts hours used,
//! applies any accrual cap, and values the remaining balance for a separation
//! payout. Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct PtoBalanceInput {
    pub employer_name: String,
    pub employee_name: String,
    /// Hours accrued each pay period.
    pub accrual_rate_hours_per_period: f64,
    /// Pay periods worked in the measurement window.
    pub periods_worked: f64,
    /// PTO hours used in the window.
    #[serde(default)]
    pub hours_used: f64,
    /// Maximum hours the balance may reach (0 = no cap).
    #[serde(default)]
    pub accrual_cap_hours: f64,
    /// Employee's hourly rate, for the payout value.
    #[serde(default)]
    pub hourly_rate_usd: f64,
    /// Hours in a workday, for the days conversion (usually 8).
    #[serde(default = "default_hours_per_day")]
    pub hours_per_day: f64,
    pub as_of_date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_hours_per_day() -> f64 {
    8.0
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct PtoBalance {
    pub title: String,
    /// Accrual rate × periods worked.
    pub hours_earned: f64,
    pub hours_used: f64,
    /// Earned − used after applying the cap, floored at 0.
    pub balance_hours: f64,
    /// True when the cap reduced the balance.
    pub cap_applied: bool,
    pub balance_days: f64,
    /// Balance × hourly rate.
    pub payout_value_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &PtoBalanceInput) -> PtoBalance {
    let earned = i.accrual_rate_hours_per_period * i.periods_worked;
    let net = earned - i.hours_used;
    let (capped, cap_applied) = if i.accrual_cap_hours > 0.0 && net > i.accrual_cap_hours {
        (i.accrual_cap_hours, true)
    } else {
        (net, false)
    };
    let balance = capped.max(0.0);
    let balance_days = if i.hours_per_day > 0.0 {
        cents(balance / i.hours_per_day)
    } else {
        0.0
    };
    let payout = cents(balance * i.hourly_rate_usd);

    let cap_line = if cap_applied {
        format!(" The balance is capped at {} hours.", cents(i.accrual_cap_hours))
    } else {
        String::new()
    };

    let calc_body = format!(
        "Earned {} hours ({} hours/period × {} periods) less {} hours used leaves a balance of {} hours ({} workdays).{} At {} per hour, the payout value is {}.",
        cents(earned),
        cents(i.accrual_rate_hours_per_period),
        cents(i.periods_worked),
        cents(i.hours_used),
        cents(balance),
        balance_days,
        cap_line,
        money(i.hourly_rate_usd),
        money(payout)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("PTO is administered under the employer's policy and the laws of the State of {}.", i.state)
    } else {
        format!("PTO is administered under the employer's policy and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Statement".into(),
            body: format!(
                "Employer: {}\nEmployee: {}\nAs of: {}",
                i.employer_name, i.employee_name, i.as_of_date
            ),
        },
        DocClause { heading: "1. Balance".into(), body: calc_body },
        DocClause {
            heading: "2. Payout".into(),
            body: format!(
                "If paid out, the {} remaining hours are valued at {} per hour for a total of {}, subject to the employer's policy and applicable law on separation.",
                cents(balance), money(i.hourly_rate_usd), money(payout)
            ),
        },
        DocClause { heading: "3. Governing Law".into(), body: governing },
        DocClause {
            heading: "Acknowledgement".into(),
            body: format!(
                "Employee: ____________________  Date: __________\n{}",
                i.employee_name
            ),
        },
    ];

    PtoBalance {
        title: "PTO Balance Statement".into(),
        hours_earned: cents(earned),
        hours_used: cents(i.hours_used),
        balance_hours: cents(balance),
        cap_applied,
        balance_days,
        payout_value_usd: payout,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> PtoBalanceInput {
        PtoBalanceInput {
            employer_name: "Acme Co".into(),
            employee_name: "Sam Worker".into(),
            accrual_rate_hours_per_period: 5.0,
            periods_worked: 20.0,
            hours_used: 30.0,
            accrual_cap_hours: 80.0,
            hourly_rate_usd: 25.0,
            hours_per_day: 8.0,
            as_of_date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn balance_and_payout() {
        let d = generate(&base());
        assert!(close(d.hours_earned, 100.0));
        assert!(close(d.balance_hours, 70.0));
        assert!(!d.cap_applied);
        assert!(close(d.balance_days, 8.75));
        assert!(close(d.payout_value_usd, 1_750.0));
    }

    #[test]
    fn cap_limits_balance() {
        let d = generate(&PtoBalanceInput { hours_used: 10.0, ..base() });
        // net 90 → capped to 80.
        assert!(d.cap_applied);
        assert!(close(d.balance_hours, 80.0));
        assert!(close(d.balance_days, 10.0));
        assert!(close(d.payout_value_usd, 2_000.0));
    }

    #[test]
    fn no_cap_uses_net() {
        let d = generate(&PtoBalanceInput { accrual_cap_hours: 0.0, ..base() });
        assert!(!d.cap_applied);
        assert!(close(d.balance_hours, 70.0));
    }

    #[test]
    fn overused_floors_at_zero() {
        let d = generate(&PtoBalanceInput { periods_worked: 5.0, hours_used: 40.0, ..base() });
        // earned 25, used 40 → net negative → 0.
        assert!(close(d.balance_hours, 0.0));
        assert!(close(d.payout_value_usd, 0.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&PtoBalanceInput { statute_citation: "Cal. Lab. Code § 227.3".into(), ..base() });
        assert_eq!(d.statutory_citation, "Cal. Lab. Code § 227.3");
        assert!(d.clauses.iter().any(|c| c.body.contains("Cal. Lab. Code § 227.3")));
    }
}
